README.md
# Perspective XSLT stylesheets

"Perspective" is a stylesheet distribution that allows you to publish structured
documents to web and print, using a custom
[DTD](https://en.m.wikipedia.org/wiki/Document_type_definition). The stylesheets
in this distribution allow the creation and transformation of XML to HTML, PDF
(using [Formatting-Objects](https://en.m.wikipedia.org/wiki/XSL_Formatting_Objects) via XSLT) and [DocBook](https://docbook.org/), a
widely used XML standard.

The importance of semantic markup is that you will have a clear separation
between style and content, the same content being used to generate layouts in
many media. Other solutions based on Wiki style markup of plain text don't
properly take into account document meta information, and they certainly don't
handle bibliographic data. This package provides that as well as defining a
compact and semantically meaningful XML that can be transformed to any other
schema. This package produces XML to the [Docbook schema](https://docbook.org/xml/5.1/),
for example. DocBook may then be used for transforming to present and future
formats such as the epub format (using the most recent versions of the DocBook
XSL stylesheets) used in the Sony Reader and Adobe Digital Editions.

The perspective-schema DTD allows the creation of articles with hyperlinks,
footnote entries, itemised lists, tables, image inclusions and citations, as
well as certain inline elements such as superscripts, bold, underline and
italic. It also handles proper quotation marks and a useful subset of foreign
characters and currency symbols. The XSLT stylesheet will insert a drop cap at
the start of the text body, process footnotes to provide forward and backward
links, and create author initials in bibliographic entries.

Perspective provides a DTD for a reduced markup language, 

      Namespace https://www.e-conomist.me.uk/xml/dtd/perspective-schema.dtd
      PUBLIC    "-//CTIPPER//DTD perspective XML V2.1//EN" 

A workflow can be defined as follows:

1. Create perspective DTD-based XML document based documents from the samples in
   xml folder.

2. Output may be produced by using the supplied Tcl/Tk scripts which use:
   - Output HTML using `stylesheets/html-perspective.xsl`
   - Output PDF by transforming to formatting-objects with `stylesheets/fo-perspective.xsl` and then using the fop renderer
   - Output Docbook XML using `stylesheets/dbk-perspective.xsl`

## Notes:

The supplied Tcl scripts use xsltproc and trang validator. [^1] You may use
GNOME libxml to process the pages. See your operating system
documentation. Libxml2 is bundled by default on macOS.

[^1]: A Tcl/Tk GUI is also provided to control document transformations after
the markup has been produced.

To produce PDF you will need a FO renderer, such as
[Apache FOP](https://xmlgraphics.apache.org/fop/) Installation is somewhat
involved see documentation particularly if you need non-standard fonts and
hyphenation.

Typical command-lines as follows though using the Tcl scripts is highly
recommended.

XHTML

    xsltproc.exe --stringparam file-name e-template 
                 --stringparam path-name "" --output ..\site\e-template.html 
    ..\stylesheets\html-perspective.xsl e-template.xml

PDF

    xsltproc.exe --stringparam file-name e-template 
                 --stringparam path-name ../ --stringparam fo-processor fop 
                 --output ..\site\pdf\e-template.fo 
    ..\stylesheets\fo-perspective.xsl e-template.xml

then

    fop -c ../config/fop.xconf -fo ..\site\pdf\e-template.fo 
        -pdf ..\site\pdf\e-template.pdf 

The variable path-name refers to the path to images/ from the output directory:
uses the Unix path syntax i.e. forward slash.

Docbook

    xsltproc.exe --output ..\docbook\e-template.xml
    ..\stylesheets\dbk-perspective.xsl e-template.xml

## LICENSE

The license for this product ("LICENSE") is fairly standard. Modification
of the software is at the user's own risk, all derivative products must give
proper attribution.

-- Copyright © 2005 Christopher Tipper. All Rights Reserved.


CHANGELOG
2016-01-25 version 2.1.12

   - Upgraded to DocBook 5.1 DTD

2013-02-19 version 2.1.11

   - added e-conomist DocBook styleheets

2012-05-04 version 2.1.10

   - added Docbook to epub functionality from docbook-xsl
   - modified Tcl scripts to provide this functionality

2012-05-01 version 2.1.10

   - converted HTML to HTML5
   - docbook stylesheets all UTF-8
   - tidied up perspective.css, now HTML5 compatible, minimal CSS3

2009-03-01 version 2.1.9

   - tweaked tx.tcl, tcmd.tcl to improve PDF processing feedback

2007-09-28 version 2.1.9

   - cleaned up unnecessary DTDs
   - cleaned up CSS and removed unnecessary tags

2007-09-05 version 2.1.9

   - updated to perspective 2.0.5, added <epigraph> element
   - updated stylesheets

2007-08-28 version 2.1.8

   - updated perspective DTD to version 2.0.3: now allows optional
     paragraphs in footnotes.
   - updated footnote in html-perspective.xsl
   - updated footnote in fo-perspective.xsl

2007-08-17 version 2.1.7

   - added hrefUtilities.bas to source folder

2007-01-26 version 2.1.6

   - replace all hyphens with non-breaking hyphens in Wd2XML

2007-01-13 version 2.1.6

   - added provision for href in drop-cap code
   - added document fragment feature for Wd2XML

2006-12-12 version 2.1.6

   - version 2.0.2 of perspective DTD -- <date> is now optional

2006-12-12 version 2.1.5

   - added horizontal rule to FO stylesheet
   - added 'date' style to paragraph template in FO
   - added a bodyx option to the font parameter list

2006-10-19 version 2.1.5

   - small change to handling of entities
   - added &gt; and &lt; fix to macro

2006-10-02 version.2.1.4

   - Moved to H1, H2, H3... headings to meet accessibility guideline
   - created test and clean shell scripts and batch files to
     check the installation

2006-09-19 version 2.1.3

   - Moved html and pdf to /site folder
   - modified Tcl/Tk scripts accordingly

2006-03-01 version 2.1.2

   - fixed <xsl:text> issue in drop-cap template -- html- and
     fo-perspective.xsl
   - added Publish menu to perspective.dot Word template

2006-02-19 version 2.1.1

   - fixed a minor enumeration issue in the DTD
   - fixed &amp; in characters() subroutine, Wd2XML.bas

2006-02-18 version 2.1.1

   - added perspective-readme.xml to demo

2006-02-16 version 2.1.1

   - amended CSS to use relative font sizes
   - corrected style bug in endnotes, html-perspective.xsl

2006-01-20 version 2.1

   - character codes in Wd2XML.bas to comply with Mac Word 2004

2005-12-10 version 2.1

   - released code under Lesser GNU Public License

2005-11-12 version 2.0.8

   - fixed stylesheets to handle use-case of table without tfoot

2005-11-09 version 2.0.8

   - fixed issue with empty table-footer in fo-perspective.xsl
   - added xep call to Tcl scripts

2005-11-08 version 2.0.8

   - A Tcl/Tk GUI created to handle document transformations

2005-10-04 version 2.0.7

   - table is left aligned if @align is not specified

2005-09-05 version 2.0.7

   - upgraded title code error handler in Wd2XML.bas. The null
     document is now guaranteed to be valid XML.
   - cleaned up xmlStop() in Wd2XML.bas
   - removed unnecessary subroutine skipTitles()

2005-08-13 version 2.0.6

   - updated table stylesheets in html-perspective.xsl and
     fo-perspective.xsl Now allows striped rows and columns, and
     theader and tbody borders. Row backgrounds can be
     uncommented from the FO stylesheet.

2005-07-31 version 2.0.6

   - updated fo-perspective.xsl table with embedded footnotes

2005-07-24 version 2.0.5

   - added table[@align] attribute to perspective DTD (now version 2.0.1)
   - updated html-perspective.xsl to allow table[@align] attribute
   - updated fo-perspective.xsl to allow table[@align] attribute
   - updated dbk-perspective.xsl to transform to table[@condition]
     attribute 

2005-07-22 version 2.0.4

   - renamed all stylesheets to be more consistent
   - updated html-perspective.xsl table with embedded footnotes

2005-07-20 version 2.0.3

   - updated fo-e-template.xsl table layout to print-style with
     borders


scripts/test.sh
#!/bin/bash
../scripts/tcmd.tcl -validate e-template
../scripts/tcmd.tcl -docbook e-template
../scripts/tcmd.tcl -dbvalidate e-template
../scripts/tcmd.tcl -html e-template
../scripts/tcmd.tcl -web e-template
../scripts/tcmd.tcl -fo e-template
../scripts/tcmd.tcl -pdf e-template
../scripts/tcmd.tcl -validate perspective
../scripts/tcmd.tcl -docbook perspective
../scripts/tcmd.tcl -dbvalidate perspective
../scripts/tcmd.tcl -html perspective
../scripts/tcmd.tcl -web perspective
../scripts/tcmd.tcl -fo perspective
../scripts/tcmd.tcl -pdf perspective
../scripts/tcmd.tcl -validate perspective-readme
../scripts/tcmd.tcl -docbook perspective-readme
../scripts/tcmd.tcl -dbvalidate perspective-readme
../scripts/tcmd.tcl -html perspective-readme
../scripts/tcmd.tcl -web perspective-readme
../scripts/tcmd.tcl -fo perspective-readme
../scripts/tcmd.tcl -pdf perspective-readme
../scripts/tcmd.tcl -validate tropical
../scripts/tcmd.tcl -docbook tropical
../scripts/tcmd.tcl -dbvalidate tropical
../scripts/tcmd.tcl -html tropical
../scripts/tcmd.tcl -web tropical
../scripts/tcmd.tcl -fo tropical
../scripts/tcmd.tcl -pdf tropical


scripts/tcmd.tcl
#!/usr/bin/env tclsh
#
#   tcmd.tcl - A script to transform XML to various formats
#   Copyright © 2005  Christopher Tipper <chris.tipper@live.co.uk>
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Lesser General Public
#   License as published by the Free Software Foundation; either
#   version 2.1 of the License, or (at your option) any later version.
#
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public
#   License along with this library; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# last updated 2016-02-09

# local path settings
set path "$::env(HOME)/Applications"

# Transform to Web

proc XsltWeb {stylesheet pathname} {
    global filename
    if [catch {exec xsltproc --stringparam file-name $filename \
                   --stringparam path-name $pathname \
                   --output ../site/$filename.html \
                   ../stylesheets/$stylesheet.xsl ../docbook/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your DocBook document: $filename is transformed to Web."
    }
}

# Transform to HTML

proc XsltHtml {stylesheet pathname} {
    global filename
    if [catch {exec xsltproc --stringparam file-name $filename \
                   --stringparam path-name $pathname \
                   --output ../site/$filename.htm \
                   ../stylesheets/$stylesheet.xsl ../xml/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your document is transformed to HTML."
    }
}

# Transform to Docbook

proc XsltDocBook {stylesheet} {
    global filename
    if [catch {exec xsltproc --output ../docbook/$filename.xml \
                   ../stylesheets/$stylesheet.xsl ../xml/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your document is transformed to DocBook DTD."
    }
}

# Transform to FO

proc XsltFO {stylesheet pathname renderer} {
    global filename
    if [catch {exec xsltproc --stringparam file-name $filename \
                   --stringparam path-name $pathname \
                   --stringparam fo-processor $renderer \
                   --stringparam dropped-caps no \
                   --output ../site/pdf/$filename.fo \
                   ../stylesheets/$stylesheet.xsl ../xml/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your document is transformed to Formatting-Objects."
    }
}

# Read and log output

proc Log {pipe} {
    global done

    if {[eof $pipe]} {
        catch {close $pipe}
        set done 1
        return
    } else {
        gets $pipe line
        puts stdout $line
    }
}

# Transform FO to PDF using FOP

proc XsltPDF {} {
    global done filename path
    # local settings on a Mac
    set path "$path/fop"

    set fofile [file normalize [file join ../site/pdf $filename.fo]]
    set pdffile [file normalize [file join ../site/pdf $filename.pdf]]
    
    if [catch {open "|$path/fop -c ../config/fop.xconf -fo {$fofile} -pdf {$pdffile} |& cat"} result] {
        puts stdout $result\n
    } else {
        fileevent $result readable [list Log $result]
        vwait done
        puts "Processing of $filename.fo to PDF is complete.\n"
    }
}

# Validate XML

proc XmlValidate {schema} {
    global filename path
    if [catch {exec java -jar $path/trang/jing.jar $schema ../xml/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your document is well-formed and validated."
    }
}

# Validate Docbook

proc DocbookValidate {schema} {
    global filename path
    if [catch {exec java -jar $path/trang/jing.jar $schema ../docbook/$filename.xml} result] {
        puts stdout $result
    } else {
        puts stdout "Your document is well-formed and validated."
    }
}

set format [lindex $argv 0]
set filename [lindex $argv 1]
if {$filename == ""} {
    puts "Usage: tclsh tcmd.tcl <option> <filename>"
    puts "Where <option> is one of -web, -html, -fo, -pdf, -docbook, -epub, "
    puts "-validate or -dbvalidate"
    puts "Please omit the filename extension in <filename>"
} else {
    switch -glob -- $format {
        -docbook {XsltDocBook dbk-perspective}
        -fo {XsltFO fo-perspective ../ fop}
        -web {XsltWeb html-dbk-perspective ""}
        -html {XsltHtml html-perspective ""}
        -pdf {XsltPDF}
        -validate {XmlValidate ../xml/dtd/perspective-schema.rng}
        -dbvalidate {DocbookValidate ../docbook/docbook-rng/docbookxi.rng}
    }
}


scripts/tx.tcl
#!/bin/sh
# the next line restarts using wish \
exec /usr/local/bin/wish "$0" ${1+"$@"}
#
#   tx.tcl - A Tk GUI to transform XML to various formats
#   Copyright © 2005  Christopher Tipper <chris.tipper@live.co.uk>
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Lesser General Public
#   License as published by the Free Software Foundation; either
#   version 2.1 of the License, or (at your option) any later version.
#
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public
#   License along with this library; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# last updated 04-05-2012

# local path settings
set path "$::env(HOME)/Applications"

wm title . Tx

# Create a frame for file-browser

frame .top -borderwidth 10
pack .top -side top -fill x

# Create the browse button

set browse [button .top.browse -text Browse -command Browse]
pack .top.browse -side right

# Create a labeled entry for the filename

label .top.l -text "File name:" -padx 0
set filename [entry .top.filename -width 20 -relief sunken]
pack .top.l -side left
pack .top.filename -side left -fill x -expand true
focus .top.browse

# Create button panel

frame .p -borderwidth 5
pack .p -side top -fill x

# Create action buttons
button .p.xhtml -text HTML -width 10 -command XsltHtml
pack .p.xhtml -side left
button .p.docbook -text DocBook -width 10 -command XsltDocBook
pack .p.docbook -side left
button .p.html -text Web -width 10 -command XsltWeb
pack .p.html -side left
button .p.valid -text Validate -width 10 -command XsltValidate
pack .p.valid -side left

# Create second button panel

frame .o -borderwidth 5
pack .o -side top -fill x

# Create more action buttons

button .o.fo -text FO -width 10 -command {XsltFO ../ fop}
pack .o.fo -side left
button .o.pdf -text PDF -width 10 -command XsltPDF
pack .o.pdf -side left

# Create log pane

frame .l
set log [text .l.log -width 60 -height 10 \
             -borderwidth 2 -relief raised -setgrid true -font {helvetica 12}\
             -yscrollcommand {.l.scroll set}]
scrollbar .l.scroll -command {.l.log yview}
pack .l.scroll -side right -fill y
pack .l.log -side left -fill both -expand true
pack .l -side top -fill both -expand true

# Browse for XML file

proc Browse {} {
    global filename
    set typelist {
        {"XML Files" {".xml"}}
    }
    set choice [tk_getOpenFile -filetypes $typelist -initialdir [file normalize [file join . ../xml]]]
    $filename delete 0 end
    $filename insert end $choice
}

# Transform to HTML

proc XsltHtml {} {
    global filename log
    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set thisdoc [file rootname [file tail $document]]
    $log delete 1.0 end
    if [catch {exec xsltproc --stringparam file-name $thisdoc \
                   --stringparam path-name "" \
                   --output [file join ../site/ $thisdoc.htm] \
                   ../stylesheets/html-perspective.xsl [file join ../xml $thisdoc.xml]} result] {
        $log insert end $result\n
    } else {
        $log insert end "Your document is transformed to HTML.\n"
    }
}

# Transform to FO

proc XsltFO {pathname renderer} {
    global filename log
    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set thisdoc [file rootname [file tail $document]]
    $log delete 1.0 end

    if [catch {exec xsltproc --stringparam file-name $thisdoc \
                   --stringparam path-name $pathname \
                   --stringparam fo-processor $renderer \
                   --stringparam dropped-caps no \
                   --output [file join ../site/pdf $thisdoc.fo] \
                   ../stylesheets/fo-perspective.xsl [file join ../xml $thisdoc.xml]} result] {
        $log insert end $result\n
    } else {
        $log insert end "Your document is transformed to Formatting-Objects.\n"
    }
}

# Read and log output

proc Log {} {
    global input log done
    if [eof $input] {
        catch {close $input}
        set done 1
        return
    } else {
        gets $input line
        $log insert end $line\n
        $log see end
    }
}

# Transform FO to PDF using FOP

proc XsltPDF {} {
    global input filename log path done

    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set thisdoc [file rootname [file tail $document]]

    set fofile [file normalize [file join ../site/pdf $thisdoc.fo]]
    set pdffile [file normalize [file join ../site/pdf $thisdoc.pdf]]

    $log delete 1.0 end
    
    if [catch {open "|$path/fop/fop -c ../config/fop.xconf -fo {$fofile} -pdf {$pdffile} |& cat"} input] {
        $log insert end $input\n
    } else {
        fileevent $input readable Log
        vwait done
        $log insert end "Processing of [file rootname [file tail [$filename get]]].fo to PDF is complete.\n"
        set done 0
    }
}

# Transform to DocBook

proc XsltDocBook {} {
    global filename log
    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set thisdoc [file rootname [file tail $document]]
    $log delete 1.0 end
    if [catch {exec xsltproc --output [file join ../docbook $thisdoc.xml] \
                   ../stylesheets/dbk-perspective.xsl [file join ../xml $thisdoc.xml]} result] {
        $log insert end $result\n
    } else {
        $log insert end "Your document is transformed to DocBook DTD.\n"
    }
}

# Validate XML

proc XsltValidate {} {
    global filename log path
    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set schema [file join "../xml/dtd" perspective-schema.rng]
    set thisdoc [file tail $document]
    $log delete 1.0 end
    if [catch {exec java -jar $path/trang/jing.jar $schema $thisdoc} result] {
        $log insert end $result\n
    } else {
        $log insert end "Your document is well-formed and validated.\n"
    }
}

# Transform to Web

proc XsltWeb {} {
    global filename log params

    set document [$filename get]
    set dirname [file dirname $document]
    cd $dirname
    set thisdoc [file rootname [file tail $document]]

    $log delete 1.0 end
    if [catch {exec xsltproc --stringparam file-name $thisdoc \
                   --stringparam path-name "" \
                   --output [file join ../site $thisdoc.html] \
                   ../stylesheets/html-dbk-perspective.xsl [file join ../docbook $thisdoc.xml]} result] {
        $log insert end $result\n
    } else {
        $log insert end "Your DocBook document is transformed to Web.\n"
    }
}
